### ISSUE TITLE
Metadata: store size of inventory in bytes

### ISSUE THREAD

**Issue by chmodsayshello** [minetest/minetest#14995]
### Problem

Within the binary data of a decompressed minetest mapblock, there is no field specifying the size of an inventory in bytes (which is the case for all other strings), see https://github.com/minetest/minetest/blob/master/doc/world_format.md#node-metadata-list .
Either the inventory string just starts, or the next metadata entry / the first node timer starts. One has to rely on string based logic (check whether the next 4 bytes are "List" (to see if there is an inventory), and then parse "blindly" until you hit bytes that as ascii are "EndInventory").

### Solutions

**Store the size of the inventory string in bytes before the actual inventory begins starting from map version 30**.

> This allows parsing code to read the next n bytes into memory, and only then deserialize the inventory. A size of 0 bytes would indicate no inventory being present.

### Alternatives

- None

**Comment by SmallJoker** [comment]
Relevant code:
https://github.com/minetest/minetest/blob/d3ca269c79aec18a8c6d938b13af5896ed978136/src/mapblock.cpp#L390-L409

What exactly is the final goal? To parse mapblock data externally? In general it should be possible to pack the entire inventory into a string with 32-bit length prefix - which requires a map version bump.

From what I can see, minetestmapper is not affected by this problem because node data is located earlier in the block data: https://github.com/minetest/minetestmapper/blob/master/BlockDecoder.cpp#L44

**Comment by chmodsayshello** [comment]
> What exactly is the final goal? To parse mapblock data externally?

Yes, I've been writing a libary to parse (and modify) minetest worlds, and I just noticed that there is no size indicator for the string here.
HOWEVER, even from minetest's perspective, wouldn't it be better to just have a size indicator so you can load the entire inventory into memory and then pass that into the deserialzation?

**Comment by appgurueu** [comment]
There might be a misunderstanding. As far as I can see, inventories are unconditionally deserialized: https://github.com/minetest/minetest/blob/d3ca269c79aec18a8c6d938b13af5896ed978136/src/nodemetadata.cpp#L76

The resulting inventory may be empty, but there is always an inventory.

Hence I don't see why you'd have to look for `List`. This is a line-based format. You just read lines until you hit a line that says `EndInventory` (which is what `Inventory::deSerialize` does). When you get a `List <name> <count>` line you initialize a new inventory list and populate it with the itemstrings given in the following lines until you hit `EndList`.

I don't think there is a potential parsing ambiguity here, even if in 60 years the metadata version has been bumped to `0x4C`.

---

I'm firmly in the "if it ain't broken, don't fix it" camp here:

* We don't really gain on code quality, since the old code has to stay around till the end of time anyways. To the contrary, we'd need to introduce a new mapblock version and a slight change in the serialization format for no real gain.
* "This allows parsing code to read the next n bytes into memory" - there is no benefit here. Well-written parsing code is usually stream-based (as is ours). There is no need to populate some kind of buffer, you just pass the mutable `istream` to `Inventory::deSerialize`, which reads the inventory by consuming part of the stream.
* Even if such a length field were to make parsing marginally simpler (which as far as I can see it doesn't here), it would complicate the serialization: Now you either need to first write to a buffer there so you know the length before you write it, or you have to do some kind of backpatching (again with a buffer), or you need to compute the length before you write. None of this is beautiful.
* This is not serialized as a string because it isn't a string, it's an inventory, a part of the serialization format.

**Comment by sfan5** [comment]
> You just read lines until you hit a line that says `EndInventory`

This literally works, see an external example [here](https://gist.github.com/sfan5/52917a0e1338bcda2176edbc416dae60#file-mapblock-parse-py-L211-L215).
